File: src/ladym/operations/decay.py
"""ACT-R base-level decay / forgetting (ARCHITECTURE.md §2 ``forget``).

Items that have not been accessed recently and have low activation are candidates for
forgetting. We never physically delete code analysis (it's derived from source), only
episodic events whose activation has fallen below a floor for a sustained period.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

from ..config import ActivationWeights
from ..operations.activation import recency_factor
from ..schema import Layer
from ..storage.store import SQLiteStore


@dataclass
class DecayReport:
    examined: int = 0
    forgotten: int = 0
    forgotten_ids: list[str] = None  # type: ignore[assignment]

    def __post_init__(self):
        if self.forgotten_ids is None:
            self.forgotten_ids = []


def decay(
    store: SQLiteStore,
    *,
    workspace: str | None = None,
    weights: ActivationWeights | None = None,
    max_age_s: float = 30 * 24 * 3600.0,
    activation_floor: float = 0.05,
    now: float | None = None,
    dry_run: bool = False,
) -> DecayReport:
    """Forget episodic events that have decayed below ``activation_floor``.

    Code analysis (L2 code_symbol/code_file), procedural playbooks, and associative edges are
    never auto-forgotten — they represent curated knowledge, not raw experience.
    """
    weights = weights or ActivationWeights()
    now = now if now is not None else time.time()
    report = DecayReport()
    ws_clause = ""
    params: list = [Layer.EPISODIC.value]
    if workspace is not None:
        ws_clause = " AND workspace = ?"
        params.append(workspace)
    rows = store.conn.execute(
        f"SELECT * FROM memories WHERE layer = ?{ws_clause}", params
    ).fetchall()
    for r in rows:
        report.examined += 1
        mem = store._row_to_memory(r)
        age = now - mem.last_access_at
        if age < max_age_s:
            continue
        # base-level activation = recency × weight + frequency × weight
        act = (
            weights.recency * recency_factor(mem.last_access_at, half_life_s=weights.recency_half_life_s, now=now)
            + weights.frequency * 0.0  # frequency log(1+n) folded in via activation_score elsewhere
        )
        if act < activation_floor:
            report.forgotten += 1
            report.forgotten_ids.append(mem.id)
            if not dry_run:
                store.delete_memory(mem.id)
    return report

File: operations/decay.go
package operations

import (
	"github.com/ProjAnvil/LadyM/config"
	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

// DecayReport reports the outcome of a decay pass.
type DecayReport struct {
	Examined     int
	Forgotten    int
	ForgottenIDs []string
}

// Decay forgets episodic events whose activation has fallen below the floor.
// Code analysis, playbooks, and edges are never auto-forgotten.
func Decay(store *storage.SQLiteStore, workspace string, weights *config.ActivationWeights, maxAgeS, activationFloor, now float64, dryRun bool) (*DecayReport, error) {
	if weights == nil {
		weights = &config.ActivationWeights{
			Similarity: 1.0, Recency: 0.3, Frequency: 0.2, Graph: 0.15,
			TypeBoost: 0.25, RecencyHalfLifeS: 7 * 24 * 3600.0,
		}
	}
	if maxAgeS == 0 {
		maxAgeS = 30 * 24 * 3600.0
	}
	if activationFloor == 0 {
		activationFloor = 0.05
	}
	if now == 0 {
		now = schema.Now()
	}
	report := &DecayReport{}
	episodes, err := store.IterMemories(workspace, string(schema.LayerEpisodic), "")
	if err != nil {
		return nil, err
	}
	for _, mem := range episodes {
		report.Examined++
		age := now - mem.LastAccessAt
		if age < maxAgeS {
			continue
		}
		act := weights.Recency*RecencyFactor(mem.LastAccessAt, weights.RecencyHalfLifeS, now) +
			weights.Frequency*0.0
		if act < activationFloor {
			report.Forgotten++
			report.ForgottenIDs = append(report.ForgottenIDs, mem.ID)
			if !dryRun {
				if err := store.DeleteMemory(mem.ID); err != nil {
					return nil, err
				}
			}
		}
	}
	return report, nil
}
